Combining grib_util with this repo

AlysonStahl-NOAA opened the pull request:
Fixes #296 
Fixes NOAA-EMC/NCEPLIBS-grib_util#332

AlysonStahl-NOAA committed:
starting with copygb util

AlysonStahl-NOAA committed:
adding cmake options from grib_util

AlysonStahl-NOAA committed:
Update developer.yml

AlysonStahl-NOAA committed:
typo

AlysonStahl-NOAA committed:
Update CMakeLists.txt

AlysonStahl-NOAA committed:
adding dependencies to workflows

AlysonStahl-NOAA committed:
check if build with w3emc

AlysonStahl-NOAA committed:
added cnvgrib utility

AlysonStahl-NOAA committed:
Update CMakeLists.txt

AlysonStahl-NOAA committed:
Update CMakeLists.txt

AlysonStahl-NOAA committed:
Update CMakeLists.txt

AlysonStahl-NOAA committed:
adding copygb2 library

AlysonStahl-NOAA committed:
typo

AlysonStahl-NOAA committed:
added the rest of the libraries

AlysonStahl-NOAA committed:
fixing some compile flags

AlysonStahl-NOAA committed:
created cmake option BUILD_UTILS

AlysonStahl-NOAA committed:
removed some redundant code

AlysonStahl-NOAA committed:
added tests and test data

AlysonStahl-NOAA committed:
fixed some paths

AlysonStahl-NOAA committed:
commenting out tests that use wgrib

AlysonStahl-NOAA committed:
missing test data

AlysonStahl-NOAA committed:
debugging CI

AlysonStahl-NOAA committed:
more debugging

AlysonStahl-NOAA committed:
disabling all util tests for debugging

AlysonStahl-NOAA committed:
debugging CI

AlysonStahl-NOAA committed:
Update tocgrib2super.F90

AlysonStahl-NOAA committed:
debugging CI

AlysonStahl-NOAA committed:
more CI debugging

AlysonStahl-NOAA committed:
turning build utils on in some workflows

AlysonStahl-NOAA committed:
Update Linux_options.yml

AlysonStahl-NOAA committed:
attempting to update spack recipe

AlysonStahl-NOAA committed:
updating spack

AlysonStahl-NOAA committed:
Update package.py

AlysonStahl-NOAA committed:
Update package.py

edwardhartnett commented:
The tests for utils should be in their own directory, test_utils. It should have it's own data subdirectory.

But we don't want duplicate data files. If the utils test needs to get data from tests/data that''s fine too.

AlysonStahl-NOAA committed:
reorganizing tests

AlysonStahl-NOAA committed:
Update CMakeLists.txt

AlysonStahl-NOAA committed:
Update run_cnvgrib_tests.sh

AlysonStahl-NOAA committed:
debugging - turning off mem checks for util tests

AlysonStahl-NOAA committed:
more debugging

AlysonStahl-NOAA committed:
removed setting

AlysonStahl-NOAA committed:
check which utils build with w3emc

AlysonStahl-NOAA committed:
Update CMakeLists.txt

AlysonStahl-NOAA committed:
changing min version of w3emc to match grib_util requirements

AlysonStahl-NOAA committed:
adding some flags to debug workflows

AlysonStahl-NOAA committed:
check build_4 and build_d before building utilities

AlysonStahl-NOAA committed:
Update CMakeLists.txt

AlysonStahl-NOAA committed:
fixed some errors in cmake

AlysonStahl-NOAA committed:
Update CMakeLists.txt

AlysonStahl-NOAA commented:
@edwardhartnett @AlexanderRichert-NOAA 

I've done pretty much all I can to get most of the checks to pass. 

The tests for the utilities fail the memory checks in the develop workflow. Is this something we can work around for now and fix down the line or should we make sure the tests pass those checks before merging?

One of the Spack checks keep failing. I did make changes to the python script myself and I'm fairly new to Spack, so someone should probably double check my work there.

I know the grib_util workflow can probably be deleted, but I want to double check before I do that.

Also, I just need to update the docs, but I'm not sure how we want to structure it. Any thoughts?

AlexanderRichert-NOAA commented:
The Spack workflow failure is because it's compiling w3emc without some extra functions that are disabled by default. In the grib-util recipe, notice it depends on `w3emc +extradeps`, so we'll need the same thing here.

AlysonStahl-NOAA committed:
Update package.py

AlysonStahl-NOAA committed:
Update package.py

AlysonStahl-NOAA committed:
turn on g2c_compare test

edwardhartnett commented:
The grib_utils CI run was the one that build grib_utils and ran its tests in order to check g2. Since we will now be building and testing them together, it's no longer needed and I will remove it...

edwardhartnett commented:
The grib_utils will fail memory tests. That's a known problem. For the purposes of this PR, just cause the developer build to not build the utilities at all. We will address this with some CI cleanup after @AlexanderRichert-NOAA merges his CI simplifying changes which will make it easier.

I will see if I can do that edit to the developer workflow...

edwardhartnett committed:
Update developer.yml

Turned off utils build, since the utils code does not pass memory checking (yet).

edwardhartnett committed:
removed grib_util CI run

edwardhartnett commented on .github/workflows/Linux_versions.yml:
```diff
@@ -30,7 +30,7 @@ jobs:
         jasper-version: [2.0.33, 3.0.5, 4.0.0]
         bacio-version: [2.4.1, 2.5.0, 2.6.0]
-        w3emc-version: [2.9.2, 2.9.3, 2.10.0]
+        w3emc-version: [2.10.0, 2.11.0]
```
Do w3emc-2.9.2 aand w3emc-2.9.3 no longer work with g2?

AlysonStahl-NOAA replied:
Yes, the grib_util library required w3emc > 2.10.0. 

It just occurred to me thought that this particular workflow does not build the tools, so I can revert the changes to make sure older versions are still getting tested.

edwardhartnett committed:
trying to fix spack CI build

edwardhartnett committed:
trying to fix spack CI build

edwardhartnett committed:
no longer trying to fix spack CI build

edwardhartnett commented:
OK, I turned off utils in the developer memory checking build. Also I removed the linux version expected builds for two versions of w3emc, but why?

@AlexanderRichert-NOAA can you help with the spack stuff? I tried to figure it out but ended up confused...

Once we get the CI problems resolved, I think we can merge...

AlysonStahl-NOAA committed:
reverting some workflow changes

AlysonStahl-NOAA committed:
Merge branch 'as_grib_util' of https://github.com/AlysonStahl-NOAA/NCEPLIBS-g2 into as_grib_util

AlexanderRichert-NOAA commented:
Sure. I think I see the issue. I'll see if I can work out a fix and I'll submit it as a PR to @AlysonStahl-NOAA's branch.

AlexanderRichert-NOAA committed:
Make various utils tests conditional on BUILD_{4,D}

AlexanderRichert-NOAA commented:
See https://github.com/AlysonStahl-NOAA/NCEPLIBS-g2/pull/1. It might be worth (at some point) adding BUILD_4/BUILD_D to the non-Spack unit tests under Linux_options.

AlysonStahl-NOAA committed:
Merge pull request #1 from AlexanderRichert-NOAA/test_fix_6jun24

Make various utils tests conditional on BUILD_{4,D}

AlysonStahl-NOAA commented:
> See [AlysonStahl-NOAA#1](https://github.com/AlysonStahl-NOAA/NCEPLIBS-g2/pull/1). It might be worth (at some point) adding BUILD_4/BUILD_D to the non-Spack unit tests under Linux_options.

Good catch, thanks for fixing that.

edwardhartnett commented:
@AlysonStahl-NOAA do you know what is happening with the failing build in CI?

Did we change the acceptable version of w3emc?

AlysonStahl-NOAA committed:
Update CMakeLists.txt

AlysonStahl-NOAA committed:
Merge branch 'as_grib_util' of https://github.com/AlysonStahl-NOAA/NCEPLIBS-g2 into as_grib_util

AlysonStahl-NOAA commented:
> @AlysonStahl-NOAA do you know what is happening with the failing build in CI?
> 
> Did we change the acceptable version of w3emc?

Yes I did change that. I changed it back to the previous requirement (2.9.0) if BUILD_UTILS is off. It will require 2.10.0 and higher if it's on.

AlexanderRichert-NOAA commented:
FWIW this looks pretty good to me-- the only thing I would add is you'll want to update ./cmake/PackageConfig.cmake.in to reflect dependencies, including adding OpenMP if enabled.

AlysonStahl-NOAA commented:
> FWIW this looks pretty good to me-- the only thing I would add is you'll want to update ./cmake/PackageConfig.cmake.in to reflect dependencies, including adding OpenMP if enabled.

Which other dependencies should be included? It looks like the only dependency it checks for currently is PNG and bacio was removed recently. It doesn't seem to check for Jasper or w3emc either.

AlexanderRichert-NOAA commented:
Are any of the changes as far as added find_package's, OpenMP, etc. related to the library itself, or is that all just for the utilities? If it's just for the utilities then I guess nothing needs to change there.

AlysonStahl-NOAA commented:
The find_package changes should be for the utilities only. In fact, I'll go ahead and only look for those packages if the utilities are enabled.

AlysonStahl-NOAA committed:
only check for utilities dependencies if build_utils

AlysonStahl-NOAA committed:
add utils path to Doxyfile.in

edwardhartnett commented:
Is this ready to merge?

AlysonStahl-NOAA commented:
> Is this ready to merge?

Yep.
